src/QMCHamiltonians/NonLocalECPComponent.cpp

```diff
@@ -597,6 +599,247 @@ NonLocalECPComponent::RealType NonLocalECPComponent::evaluateOneWithForces(Parti
+  {
+    W.makeMove(iel, deltaV[j], false); //Update distance tables.
+    psi.getRowM(W, iel, phi_row);
+    W.rejectMove(iel);
+
+    RealType zz = dot(dr, rrotsgrid_m[j]) * rinv;
+    // Forming the Legendre polynomials
+    lpol[0]           = cone;
+    RealType lpolprev = czero;
+    for (int l = 0; l < lmax; l++)
+    {
+      //Not a big difference
+      //lpol[l+1]=(2*l+1)*zz*lpol[l]-l*lpolprev;
+      //lpol[l+1]/=(l+1);
+      lpol[l + 1] = Lfactor1[l] * zz * lpol[l] - l * lpolprev;
+      lpol[l + 1] *= Lfactor2[l];
+      lpolprev = lpol[l];
+    }
+
+    ValueType lsum = 0.0;
+    for (int l = 0; l < nchannel; l++)
+    {
+      temp_row = (vrad[l] * lpol[angpp_m[l]] * sgridweight_m[j]) * phi_row;
+      for (int iorb = 0; iorb < numOrbs; iorb++)
+        B[sid][thisIndex][iorb] += temp_row[iorb];
+    }
+  }
+}
+
+void NonLocalECPComponent::evaluateOneBodyOpMatrixdRContribution(ParticleSet& W,
+                                                                 const ParticleSet& ions,
+                                                                 const int iat,
+                                                                 const int iat_src,
+                                                                 const TWFFastDerivWrapper& psi,
+                                                                 const int iel,
+                                                                 const RealType r,
+                                                                 const PosType& dr,
+                                                                 std::vector<std::vector<ValueMatrix>>& dB)
+{
+  using ValueVector = SPOSet::ValueVector;
+  using GradVector  = SPOSet::GradVector;
+  constexpr RealType czero(0);
+  constexpr RealType cone(1);
+
+  //We check that our quadrature grid is valid.  Namely, that all points lie on the unit sphere.
+  //We check this by seeing if |r|^2 = 1 to machine precision.
+  for (int j = 0; j < nknot; j++)
+    assert(std::abs(std::sqrt(dot(rrotsgrid_m[j], rrotsgrid_m[j])) - 1) <
+           100 * std::numeric_limits<RealType>::epsilon());
+
+  for (int j = 0; j < nknot; j++)
+    deltaV[j] = r * rrotsgrid_m[j] - dr;
+
+  // This is just a temporary variable to dump d2/dr2 into for spline evaluation.
+  RealType secondderiv(0);
+
+  const RealType rinv = cone / r;
+
+  // Compute radial potential and its derivative times (2l+1)
+  for (int ip = 0; ip < nchannel; ip++)
+  {
+    //fun fact.  NLPComponent stores v(r) as v(r), and not as r*v(r) like in other places.
+    vrad[ip]  = nlpp_m[ip]->splint(r, dvrad[ip], secondderiv) * wgt_angpp_m[ip];
+    vgrad[ip] = dvrad[ip] * dr * wgt_angpp_m[ip] * rinv;
+  }
+
+  IndexType gid        = W.getGroupID(iel);
+  IndexType sid        = psi.getTWFGroupIndex(gid);
+  IndexType thisEIndex = iel - W.first(gid);
+  IndexType numptcls   = W.last(gid) - W.first(gid);
+  IndexType norbs      = psi.numOrbitals(sid);
+  SPOSet* spo          = psi.getSPOSet(sid);
```

Comment:
Whenever possible, I tries to avoid pointers.
```
SPOSet& spo(*psi.getSPOSet(sid));
```
but this just more of my taste, I will let you decide.

Reply:
I'd prefer to avoid pointers too.  Done.